File: src/main/java/it/alian/gun/mesmerize/compat/EntityName.java
package it.alian.gun.mesmerize.compat;

import org.bukkit.Bukkit;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;

public abstract class EntityName {

    private static EntityName impl;

    public static String get(Entity entity) {
        return impl.getName(entity);
    }

    public static void init() {
        if (Bukkit.getServer().getClass().getName().split("\\.")[3].equals("v1_7_R4"))
            impl = new Impl_1_7();
        else
            impl = new Impl_1_8();
    }

    public abstract String getName(Entity entity);

    private static class Impl_1_7 extends EntityName {

        @Override
        public String getName(Entity entity) {
            if (entity instanceof Player)
                return ((Player) entity).getDisplayName();
            else
                return entity.getType().getName();
        }
    }

    private static class Impl_1_8 extends EntityName {

        @Override
        public String getName(Entity entity) {
            if (entity instanceof Player)
                return ((Player) entity).getDisplayName();
            else
                return entity.getName();
        }
    }

}

File: src/main/scala/it/alian/gun/mesmerize/compat/EntityName.scala
package it.alian.gun.mesmerize.compat

import org.bukkit.Bukkit
import org.bukkit.entity.{Entity, Player}

object EntityName {
  private val impl = if (Bukkit.getServer.getClass.getName.split("\\.")(3) == "v1_7_R4") new Impl_1_7 else new Impl_1_8

  def get(entity: Entity): String = impl(entity)

  private class Impl_1_7 extends (Entity => String) {
    override def apply(entity: Entity): String = entity match {
      case player: Player => player.getDisplayName
      case _ => entity.getType.getName
    }
  }

  private class Impl_1_8 extends (Entity => String) {
    override def apply(entity: Entity): String = entity match {
      case player: Player => player.getDisplayName
      case _ => entity.getName
    }
  }

}